### PR TITLE
Fix to domain for bounding-box (support having only a single element …

### PR DESCRIPTION
**Pull request by geoalgo**
…and provide support for restricting finite search spaces).

*Issue #, if available:*

*Description of changes:* pull changes from automl experiment branch that affects common components (this is the last one the other PR will be just new schedulers and experiment scripts)

* support finite range with only single elements
* support restricting finite range to smaller intervals
* support not loading tuner when accessing results (much faster)
* use metadata instead of inspecting the tuner for mode and metrics
* allow to load metadata only of experiments and filter by name (much faster than loading all experiments results)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: syne_tune/config_space.py [hunk 1]
```diff
@@ -967,3 +973,38 @@ def from_dict(d: Dict) -> Domain:
...
+    """
+    Restricts a numerical domain to be in the range [lower, upper]
+    :return:
+    """
+    assert hasattr(numerical_domain, "lower") and hasattr(numerical_domain, "upper")
+    assert lower <= upper
+    if not isinstance(numerical_domain, FiniteRange):
+        # domain is numerical, set new lower and upper ranges with bounding-box values
+        new_domain_dict = to_dict(numerical_domain)
+        new_domain_dict['domain_kwargs']['lower'] = lower
```
Comment (mseeger) [inline syne_tune/config_space.py:988]: Cast upper, lower to Domain type?

File: syne_tune/config_space.py [hunk 2]
```diff
@@ -967,3 +973,38 @@ def from_dict(d: Dict) -> Domain:
...
+        new_domain_dict['domain_kwargs']['upper'] = upper
+        return from_dict(new_domain_dict)
+    else:
+        values = numerical_domain._values
+        assert lower < max(numerical_domain._values)
```
Comment (mseeger) [inline syne_tune/config_space.py:993]: Again, first cast upper, lower to domain type. Important for int.

File: syne_tune/config_space.py [hunk 3]
```diff
@@ -967,3 +973,38 @@ def from_dict(d: Dict) -> Domain:
...
+        new_lower = values[i]
+
+        j = len(values) - 1
+        while upper < values[j] and i < j:
+            j -= 1
+        new_upper = values[j]
+        return FiniteRange(
+            lower=new_lower,
+            upper=new_upper,
+            size=len(values) - i - (len(values) - j - 1),
```
Comment (mseeger) [inline syne_tune/config_space.py:1007]: j - i + 1

File: syne_tune/config_space.py [hunk 4]
```diff
@@ -612,38 +612,44 @@ def __init__(self, lower: float, upper: float, size: int,
...
             self._lower_internal = lower
-            self._step_internal = (upper - lower) / (size - 1)
+            self._step_internal = (upper - lower) / (size - 1) if size > 1 else 0
         else:
             self._lower_internal = np.log(lower)
             upper_internal = np.log(upper)
-            self._step_internal = \
-                (upper_internal - self._lower_internal) / (size - 1)
+            self._step_internal = (upper_internal - self._lower_internal) / (size - 1) if size > 1 else 0
+        self._values = [self._map_from_int(x) for x in range(self.size)]
```
Comment (mseeger) [inline syne_tune/config_space.py:620]: Could you make this a property, say self.values, and then use it in the function below?

File: syne_tune/experiments.py [hunk 5]
```diff
@@ -144,13 +146,15 @@ def load_experiment(
...
         results = None
-    try:
-        tuner = Tuner.load(path)
-    except FileNotFoundError:
-        tuner = None
-    except Exception:
+    if load_tuner:
+        try:
+            tuner = Tuner.load(path)
+        except FileNotFoundError:
```
Comment (mseeger) [inline syne_tune/experiments.py:152]: Is this needed?

**Review by mseeger — COMMENTED** [review]
I am missing hp_ranges_impl, it had the same bug as config_space.

File: syne_tune/config_space.py (hunk 1 above)
Comment (geoalgo) [reply syne_tune/config_space.py:988]: Done.

File: syne_tune/config_space.py (hunk 3 above)
Comment (geoalgo) [reply syne_tune/config_space.py:1007]: Done.

File: syne_tune/config_space.py (hunk 4 above)
Comment (geoalgo) [reply syne_tune/config_space.py:620]: I chose this way as it allows to cache the values when sampling.

File: syne_tune/experiments.py (hunk 5 above)
Comment (geoalgo) [reply syne_tune/experiments.py:152]: There may be cases where you cannot load the tuner but still want the results.

**Comment by geoalgo** [comment]
> I am missing hp_ranges_impl, it had the same bug as config_space.

Sorry, it is not clear what you mean by that.

File: syne_tune/config_space.py (hunk 4 above)
Comment (mseeger) [reply syne_tune/config_space.py:620]: Yes, I like it, but maybe we should have a property, otherwise you access something internal below

**Comment by mseeger** [comment]
What I mean: There is also .../bayesopt/datatypes/hp_ranges_impl.py, where you have HyperparameterRangeFiniteRange, and I think you want to add your change there as well, so it also supports finrange of size 1.

**Comment by mseeger** [comment]
Right now, this would not work for BO methods, because they use hp_ranges_impl and would fail there.

File: syne_tune/config_space.py (hunk 4 above)
Comment (geoalgo) [reply syne_tune/config_space.py:620]: I am not sure I follow, `_values` is accessed only by the class right?

Comment (geoalgo) [reply syne_tune/config_space.py:620]: Oh I see, you mean in the restrict domain method, I will add a property as well.

**Review by mseeger — APPROVED** [review]

**Pull request merged**